WIP 10 sale line option

bealdav opened the pull request:
This module provide an options system for sale order line

![bom](https://user-images.githubusercontent.com/1853434/38202302-88e6d752-369b-11e8-962d-0588e3e8625c.png)

![sale_line](https://user-images.githubusercontent.com/1853434/38202307-8e7f4410-369b-11e8-8e57-53c80cda0013.png)

bealdav committed:
ADD sale_line_option

bealdav committed:
PORT v10

bealdav committed:
FIX security, demo, sale.py

bealdav committed:
FIX unit test

bealdav committed:
FIX demo, complete view

bealdav commented:
weird result flag on run bot because:


+=======================================
|  Tests summary:
|---------------------------------------
| test_server.py              [1;32mSuccess[0;m
+=======================================

bealdav committed:
REFactor remove option.type, add max_qty

bealdav committed:
FIX play onchange in sale line

bealdav committed:
FIX manufactring order creation

bealdav committed:
IMP sale view

JayVora-SerpentCS commented on sale_line_option/README.rst:
```diff
@@ -0,0 +1,104 @@
+
+.. image:: https://odoo-community.org/website/image/ir.attachment/5784_f2813bd/datas
+   :alt: Try me on Runbot
+   :target: https://runbot.odoo-community.org/runbot/167/8.0
```
10.0

JayVora-SerpentCS commented on sale_line_option/__manifest__.py:
```diff
@@ -0,0 +1,29 @@
+    'name': 'Sale Line Option',
+    'version': '10.0.0.0.1',
+    'author': 'Akretion, Odoo Community Association (OCA)',
+    'website': 'www.akretion.com',
```
website -https://github.com/OCA/sale-workflow

nikul-serpentcs commented on sale_line_option/__manifest__.py:
```diff
@@ -0,0 +1,29 @@
+        'views/mrp_view.xml',
+        'views/stock_view.xml',
+        'views/install.xml',
+        'security/ir.model.access.csv',
```
@bealdav `security` file always first.

nikul-serpentcs commented on sale_line_option/models/sale.py:
```diff
@@ -0,0 +1,188 @@
+            if record.bom_line_id and record.opt_max_qty < record.qty:
+                max_val = record.qty
+                record.qty = record.opt_max_qty
+                return {'warning': {
```
here use `UserError`

nikul-serpentcs commented on sale_line_option/tests/test_change_option.py:
```diff
@@ -0,0 +1,68 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+import logging
+from openerp.tests.common import TransactionCase
```
`openerp` replace `odoo`

nikul-serpentcs reviewed (commented):
Improve Code

mourad-ehm committed:
[FIX] At the moment, the client does not support updates on a *2many field

hparfr committed:
Use product_qty instead of option_qty.

And set "1" as default value for Max Quantity.

hparfr committed:
Increment version because of the breaking change

hparfr committed:
Rework qty min / max / default

hparfr committed:
Update demo.xml

hparfr committed:
Update tests

hparfr committed:
Fix price_get

Price_get returns the price per unit for a given qty.

mourad-ehm committed:
[FIX] dont' replace option values by default values in create sale line

mourad-ehm committed:
[FIX] add sale_add_missing_fields_get_onchange_fields  hook for option_ids field

sebastienbeau committed:
[REF] refactor code

- remove base_unit_price not use in our project in order to simplify
the code
- try to fix issue related to the broken one2many issue
- remove field product_ids on sale.order.line.option that broke
randomly linked to onchange issue
- remove useless dependency on onchange helper
- add the concept of "with_option" on bom to define which bom are
"configurable"
- improve bom UI by adding the field optionnal_bom_line_ids
- remove deprecated code

sebastienbeau committed:
[REF] continue refactoring

- add test on sale order validation
- use computed field stored for bom_line_id on option

sebastienbeau committed:
[REF] remove useless code

The name have been introduced in the first v8 version because we where
using the bom_line_id to select the option.
As now we use the product_id is not necessary anymore

sebastienbeau committed:
[REF] USe bom_find

Avoid logic replication by using bom_find for getting the bom and line
information, this avoid hack on search/name_search

sebastienbeau committed:
[REF] add sql contraint and move comment

Note I have remove the constraint opt_min_qty <= opt_max_qty as already
check with the two other constraint

sebastienbeau committed:
[IMP] improve UI if

improve UI is quantity is not correct (the line is red)
add test on quantity
add test on price and fix computation

sebastienbeau committed:
[FIX] fix pep8

sebastienbeau committed:
[IMP] move broken sale_triple_discount in a separated test

sebastienbeau committed:
[FIX] fix travis

sebastienbeau committed:
[FIX] remove option from tree view that break randomly

rousseldenis commented:
@bealdav  @sebastienbeau Is this still applicable ?

bealdav commented:
I think not, but I let @sebastienbeau answer.

sebastienbeau commented:
Hi we totally have refactored this module, from now the project is here : https://github.com/akretion/sale-configurator
When it will be ready in V14 we will propose the repo into the OCA
